### PR TITLE
1096 fetch data from prison register api and prisoner search api

### PR DESCRIPTION
**Pull request by anandelangovan1**
## Context

[](https://trello.com/c/t8Xd0A3H/1096-add-fields-from-external-apis-to-caselists-endpoint-m)

## Changes in this PR
* ReferralSummary to return values fetched from prison search and prisoner search api
* We use wiremock to mock the response in our tests

## Release checklist

[Release process documentation](../doc/how-to/perform-a-release.md)

As part of our continuous deployment strategy we must ensure that this work is
ready to be released once merged.

### Pre-merge

- [ ] There are changes required to the Accredited Programmes UI for this change to work...
  - [ ] ... and they been released to production already

### Post-merge

- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-preprod-environment) release to preprod
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-production-environment) release to prod

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/client/prisonRegisterApi/PrisonRegisterService.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.http.HttpStatus
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+import uk.gov.justice.digital.hmpps.hmppsaccreditedprogrammesapi.client.ClientResult
+import uk.gov.justice.digital.hmpps.hmppsaccreditedprogrammesapi.results.AuthorisableActionResult
+
+@Service
+@Transactional
+class PrisonRegisterService
```
Comment (Gweaton) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/client/prisonRegisterApi/PrisonRegisterService.kt:12]: Can we call this (and the file) `PrisonService`? On the UI, we tend to have a pattern of the client being named after the API, but the service being named after the entity we're dealing with internally so I think this would be a good thing to keep consistent.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/client/prisonSearchApi/PrisonerSearchApiService.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.http.HttpStatus
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+import uk.gov.justice.digital.hmpps.hmppsaccreditedprogrammesapi.client.ClientResult
+import uk.gov.justice.digital.hmpps.hmppsaccreditedprogrammesapi.client.prisonSearchApi.model.Prisoner
+import uk.gov.justice.digital.hmpps.hmppsaccreditedprogrammesapi.results.AuthorisableActionResult
+
+@Service
+@Transactional
+class PrisonerSearchApiService
```
Comment (Gweaton) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/client/prisonSearchApi/PrisonerSearchApiService.kt:12]: See above re: `PrisonService`- let's call this `PrisonerService` to keep it agnostic from the API it's calling.

**Review by Gweaton — COMMENTED** [review]
Nice one for getting this over the line - have left a few comments here.

For committing this, I'd split it the following way:
- Add prison register & db to docker compose YML for running locally
- Add PrisonService
- Add PrisonerService
- Add Prisoners and Prisons to getReferralsByOrganisationId response

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/restapi/transformer/ReferralTransformers.kt [hunk 3]
```diff
@@ -47,20 +49,48 @@ fun ReferralUpdate.toApi() = ApiReferralUpdate(
...
         prisonNumber = firstProjection.prisonNumber,
+        organisationId = organisationId,
+        prisonName = allPrisonDetails[organisationId].orEmpty(),
+        prisonerName = getPrisonerName(firstProjection, prisonersDetails),
+        sentenceDate = getSentenceDate(firstProjection, prisonersDetails),
       )
     }
 }
+
+private fun getSentenceDate(referralSummaryProjection: ReferralSummaryProjection, prisonersDetails: Map<String?, List<Prisoner>>): SentenceDate? {
```
[the remark this answers is not in the document]
Comment (anandelangovan1) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/restapi/transformer/ReferralTransformers.kt:77]: The naming `SentenceDates` usually signifies a list. Changed the object to Sentence

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/unit/restapi/controller/ReferralControllerTest.kt [hunk 4]
```diff
@@ -445,7 +482,7 @@ constructor(
     }
 
     every { referralService.getReferralsByOrganisationId(organisationId, pageable, null, null, null) } returns
-      PageImpl(projectionsForFirstReferral.toApi() + projectionsForSecondReferral.toApi(), pageable, 2L)
+      PageImpl(projectionsForFirstReferral.toApi(prisoners, prisons, organisationId) + projectionsForSecondReferral.toApi(prisoners, prisons, organisationId), pageable, 2L)
```
Comment (nsmj63-dxw) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/unit/restapi/controller/ReferralControllerTest.kt:485]: See how `.toApi()` is now more complicated than a simple transformation? Looks more like a builder or something. Possibly hard to parse for newcomers to the project.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/unit/restapi/controller/ReferralControllerTest.kt [hunk 5]
```diff
@@ -458,7 +495,7 @@ constructor(
     }.andReturn()
 
     val jsonResponse = mvcResult.response.contentAsString
-    val paginatedReferralSummary = jacksonObjectMapper()
+    val paginatedReferralSummary = objectMapper
```
Comment (nsmj63-dxw) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/unit/restapi/controller/ReferralControllerTest.kt:498]: Always use the Kotlin-implementation `jacksonObjectMapper()` where possible, as it's implemented as a singleton and runs much more efficiently than manually rolling it.

**Review by nsmj63-dxw — CHANGES_REQUESTED** [review]
A few stylistic changes requested, but it's 95% awesome.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/client/prisonRegisterApi/PrisonClient.kt [hunk 6]
```diff
@@ -6,15 +6,18 @@ import org.springframework.stereotype.Component
 import org.springframework.web.reactive.function.client.WebClient
 import uk.gov.justice.digital.hmpps.hmppsaccreditedprogrammesapi.client.BaseHMPPSClient
 import uk.gov.justice.digital.hmpps.hmppsaccreditedprogrammesapi.client.prisonRegisterApi.model.PrisonDetails
-import java.util.*
 
 @Component
-class PrisonRegisterApiClient(
+class PrisonClient(
   @Qualifier("prisonRegisterApiWebClient")
```
[the remark this answers is not in the document]
Comment (anandelangovan1) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/client/prisonRegisterApi/PrisonClient.kt:12]: I did think about this, but technically the service is called `prisonRegisterApi` but for ease of readability we call it the PrisonClient

We also have another external service called the `PrisonApi` So I think I should go back to calling this PrisonRegisterApiClient

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/restapi/transformer/ReferralTransformers.kt [hunk 7]
```diff
@@ -47,20 +49,48 @@ fun ReferralUpdate.toApi() = ApiReferralUpdate(
...
         prisonNumber = firstProjection.prisonNumber,
+        organisationId = organisationId,
+        prisonName = allPrisonDetails[organisationId].orEmpty(),
+        prisonerName = getPrisonerName(firstProjection, prisonersDetails),
+        sentenceDate = getSentence(firstProjection, prisonersDetails),
       )
     }
 }
+
+private fun getSentence(referralSummaryProjection: ReferralSummaryProjection, prisonersDetails: Map<String?, List<Prisoner>>): Sentence? {
```
[the remark this answers is not in the document]
Comment (anandelangovan1) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/restapi/transformer/ReferralTransformers.kt:77]: Wouldn't this be mutating the value of the variables?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/client/prisonRegisterApi/PrisonClient.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (anandelangovan1) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/client/prisonRegisterApi/PrisonClient.kt]: Updated

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/restapi/transformer/ReferralTransformers.kt [hunk 8]
```diff
@@ -47,20 +49,48 @@ fun ReferralUpdate.toApi() = ApiReferralUpdate(
   oasysConfirmed = oasysConfirmed,
   hasReviewedProgrammeHistory = hasReviewedProgrammeHistory,
 )
-
-fun List<ReferralSummaryProjection>.toApi(): List<ApiReferralSummary> {
+fun List<ReferralSummaryProjection>.toApi(
```
[the remark this answers is not in the document]
Comment (nsmj63-dxw) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/restapi/transformer/ReferralTransformers.kt:52]: As we've spoken about this and decided to keep things as they are, this is resolved.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/restapi/transformer/ReferralTransformers.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (nsmj63-dxw) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/restapi/transformer/ReferralTransformers.kt:77]: Now spoken about and resolved.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/common/config/JwtAuthHelper.kt [hunk 9]
```diff
@@ -35,7 +66,7 @@ class JwtAuthHelper {
       put("user_name", auth?.name ?: CLIENT_USERNAME)
       put("authorities", auth?.authorities?.map { it.authority } ?: listOf<String>())
       put("scope", listOf<String>())
-      put("client_id", "hmpps-accredited-programmes-ui")
+      put("client_id", "integration-test-client-id")
```
Comment (nsmj63-dxw) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/common/config/JwtAuthHelper.kt:69]: Not sure if this change might cause Pact failures?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/common/config/JwtAuthHelper.kt [hunk 10]
```diff
@@ -22,6 +23,36 @@ class JwtAuthHelper {
...
+  internal fun createClientCredentialsJwt(
+    username: String? = null,
+    scope: List<String>? = listOf(),
+    roles: List<String>? = listOf(),
+    authSource: String = if (username == null) "none" else "nomis",
+    expiryTime: Duration = Duration.ofHours(1),
+    jwtId: String = UUID.randomUUID().toString(),
+  ): String =
+    mutableMapOf<String, Any>()
+      .also { it["user_name"] = username ?: "integration-test-client-id" }
```
[the remark this answers is not in the document]
Comment (nsmj63-dxw) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/common/config/JwtAuthHelper.kt:40]: This still ought to be moved, I think.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/integration/ReferralIntegrationTest.kt [hunk 11]
```diff
@@ -35,6 +41,13 @@ import java.util.UUID
 @Import(JwtAuthHelper::class)
 class ReferralIntegrationTest : IntegrationTestBase() {
 
+  lateinit var bearerToken: String
+
+  @BeforeEach
+  fun setup() {
+    bearerToken = jwtAuthHelper.bearerToken()
```
[the remark this answers is not in the document]
Comment (nsmj63-dxw) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/integration/ReferralIntegrationTest.kt:48]: Let's remove this and keep the current usage as-is.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/integration/ReferralIntegrationTest.kt [hunk 12]
```diff
@@ -235,7 +283,7 @@ class ReferralIntegrationTest : IntegrationTestBase() {
     webTestClient
       .post()
       .uri("/referrals")
-      .header(HttpHeaders.AUTHORIZATION, jwtAuthHelper.bearerToken())
+      .header(HttpHeaders.AUTHORIZATION, bearerToken)
```
Comment (nsmj63-dxw) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/integration/ReferralIntegrationTest.kt:286]: Let's stick to the previous implementation.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/integration/ReferralIntegrationTest.kt [hunk 13]
```diff
@@ -254,7 +302,7 @@ class ReferralIntegrationTest : IntegrationTestBase() {
     webTestClient
       .get()
       .uri("/referrals/$createdReferralId")
-      .header(HttpHeaders.AUTHORIZATION, jwtAuthHelper.bearerToken())
+      .header(HttpHeaders.AUTHORIZATION, bearerToken)
```
Comment (nsmj63-dxw) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/integration/ReferralIntegrationTest.kt:305]: Let's stick to the previous implementation.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/integration/ReferralIntegrationTest.kt [hunk 14]
```diff
@@ -293,7 +341,7 @@ class ReferralIntegrationTest : IntegrationTestBase() {
     return webTestClient
       .get()
       .uri(uriBuilder.toUriString())
-      .header(HttpHeaders.AUTHORIZATION, jwtAuthHelper.bearerToken())
+      .header(HttpHeaders.AUTHORIZATION, bearerToken)
```
Comment (nsmj63-dxw) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/integration/ReferralIntegrationTest.kt:344]: Let's stick to the previous implementation.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/unit/restapi/controller/ReferralControllerTest.kt (hunk 4 above)
Comment (nsmj63-dxw) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/unit/restapi/controller/ReferralControllerTest.kt:480]: Spoken about and resolved.

**Review by nsmj63-dxw — DISMISSED** [review]
99% awesome now. Tiny bit of cleanup and we're good to go.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/client/prisonSearchApi/PrisonerSearchApiServiceTest.kt [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+import uk.gov.justice.digital.hmpps.hmppsaccreditedprogrammesapi.client.ClientResult
+import uk.gov.justice.digital.hmpps.hmppsaccreditedprogrammesapi.client.prisonSearchApi.model.Prisoner
+
+private const val PRISON_NUMBER_1 = "1"
+
+private const val PRISON_NUMBER_2 = "2"
+
+class PrisonerSearchApiServiceTest {
+
+  private val prisonSearchApiClient = mockk<PrisonerSearchApiClient>()
```
Comment (Gweaton) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/client/prisonSearchApi/PrisonerSearchApiServiceTest.kt:20]: let's call this variable `prisonerSearchApiClient` as it's easily confused with searching for a `prison` otherwise!

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/client/prisonSearchApi/PrisonerSearchApiServiceTest.kt [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+    val result = service.getPrisoners(listOf(PRISON_NUMBER_1, PRISON_NUMBER_2))
+
+    val expectedMap = mapOf(PRISON_NUMBER_1 to listOf(prisoner1), PRISON_NUMBER_2 to listOf(prisoner2))
+
+    assertEquals(expectedMap, result)
+  }
+
+  @Test
+  fun `getAllPrisons call returns empty list when there is a failure fetching prison details`() {
```
Comment (Gweaton) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/client/prisonSearchApi/PrisonerSearchApiServiceTest.kt:40]: `getAllPrisoners` ... `prisoner details`?

**Review by Gweaton — COMMENTED** [review]
Looking good. Thanks for making those changes, just one last thing - probably a copy and paste issue!

**Review by Gweaton — APPROVED** [review]
